### PR TITLE
Use our own PyJWKClient instead of the one provided by PyJWT

### PR DESCRIPTION
**Pull request by marcospri**
Our version uses requests instead of urllib which provides an user agent
by default but also allows for further customizations down the line.

This is motivated by school having problems with this. I reckon it makes sense to have control over this in any case.

# Testing

- On `main` test opening an LTI1.3 assignment with a modified JWK enpoint

```tox -qe dockercompose -- exec postgres psql -U postgres -c "update lti_registration set key_set_url = 'https://elearning.aua.am/mod/lti/certs.php' where id = 3"```

- Open https://hypothesis.instructure.com/courses/319/assignments/3308

Fails with

```
web (stderr)         |   File "/home/marcos/.pyenv/versions/3.8.12/lib/python3.8/urllib/request.py", line 649, in http_error_default
web (stderr)         |     raise HTTPError(req.full_url, code, msg, hdrs, fp)
web (stderr)         | urllib.error.HTTPError: HTTP Error 403: Forbidden
```

- Open https://hypothesis.instructure.com/courses/319/assignments/3308 again on `jwt-requests`

Now it fails with:

```2022-08-19 11:26:58,809 DEBUG [lms.services.jwt:98][MainThread] Invalid JWT. Unable to find a signing key that matches: "2022-07-01T00:00:03Z_0bbb730e-14aa-4420-9efa-f03583c8b777"```

which makes sense because we are using the JWKs from another LMS but the JWK was read succesful.

- Check a successful launch in canvas, go back to the canvas JWKs

```
tox -qe dockercompose -- exec postgres psql -U postgres -c "update lti_registration set key_set_url = 'https://canvas.instructure.com/api/lti/security/jwks' where id = 3"
```

An launch https://hypothesis.instructure.com/courses/319/assignments/3308

which should succeed now.

### CODE REVIEW

File: lms/services/jwt.py [hunk 1]
```diff
@@ -115,7 +116,21 @@ def _get_jwk_client(jwk_url: str):
...
+
+
+class _RequestsPyJWKClient(jwt.PyJWKClient):
+    """
+    Version of PyJWKClient which uses requests to gather JWKs.
+
+    Having our own class and using request allows for easier customization.
+    """
+
+    def fetch_data(self):
```
Comment (marcospri) [inline lms/services/jwt.py:129]: Original version at:

https://github.com/jpadilla/pyjwt/blob/a988e1a11e5abb5869dd641f3f4f6a5bb4e70fdf/jwt/jwks_client.py#L20

We could have just add the headers and keep using urllib but using requests feels more "idiomatic" in our codebase.

File: tests/unit/lms/services/jwt_test.py [hunk 2]
```diff
@@ -152,6 +164,22 @@ def svc(self, lti_registration_service, rsa_key_service):
         return svc
 
 
+class Test_RequestsPyJWKClient:
```
Comment (marcospri) [inline tests/unit/lms/services/jwt_test.py:167]: I think `_RequestsPyJWKClient` should be private but also tested independently.

Here we test our changes to the class and the test that use the mocked version tests how it all plays together.

**Review by jon-betts — APPROVED** [review]
Yep this works as described. I like splitting out and testing a sub-chunk of the problem. I think this is a good enough reason to split off a helper class sometimes and is something we should consider more for big and complicated classes with repeated chunks of behavior.

File: lms/services/jwt.py [hunk 3]
```diff
@@ -115,7 +116,21 @@ def _get_jwk_client(jwk_url: str):
...
+        # We found that some Moodle instances return 403s
+        # on request without an User-Agent.
+        with requests.get(self.uri, headers={"User-Agent": "requests"}) as response:
+            return response.json()
```
[the remark this answers is not in the document]
Comment (marcospri) [reply lms/services/jwt.py:133]: > I think requests sends it's own user agent by default?

It does this is a bit more explicit about what's it's doing. I wrote a version of that and tried to assert that there was indeed an user agent and it become a bit too complex on the testing side (`request.utils.default_heaaders()`)

**Pull request merged**